core/ppu.cpp

```diff
@@ -145,24 +230,101 @@ void PPU::renderTiles() {
 
   // LCDC Bit 4	BG and Window tile data area	0=8800-97FF, 1=8000-8FFF
   u16 tileData = tileDataArea();
+  bool unsig = checkBit(get_lcdc(), 4);
+
+  // Check if window's Y position is within the current scanline and window is enabled
+  // Tetris doesn't use a window, so this should be just set to the bgTileMapAreaa()
+  u16 tileMap;
+  bool winEnabled = (isWindowEnabled() && get_wy() <= get_ly());
+  if (!winEnabled) {
+    tileMap = bgTileMapArea();
+  } else {
+    tileMap = windowTileMapArea();
+  }
 
-  // LCDC Bit 3	BG tile map area	0=9800-9BFF, 1=9C00-9FFF
-  u16 bgMap = bgTileMapArea();
+  u8 currentLine = get_ly();
 
-  // LCDC Bit 5	Window enable	0=Off, 1=On
-  // Check if window's Y position is within the current scanline
-  // LCDC Bit 6	Window tile map area	0=9800-9BFF, 1=9C00-9FFF
-  u16 winMap;
-  if (isWindowEnabled() && get_wy() <= get_ly()) {
-    winMap = windowTileMapArea();
+  // calculate current row of tiles we are on
+  u8 yPos = 0;
+  if (!winEnabled) {
+    yPos = scrollY + currentLine;
+  } else {
+    yPos = currentLine - windowY;
   }
 
+  // calculate which row of pixels in the above tile we are on (32 tiles vertical, 8 pixels per tile)
+  u16 currPixelRow = ((u8)(yPos/8)*32);
+
   // draw current line of pixels
-  for (int i = 0; i < LCD_HEIGHT; i++) {
+  for (int i = 0; i < LCD_WIDTH; i++) {
+    u8 xPos = i + scrollX;
+
+    // same as above for x offset (if window is enabled)
+    if (winEnabled) {
+      if (i >= windowX) {
+        xPos = i - windowX;
+      }
+    }
+    
+    u16 currPixelCol = (xPos/8);
+    u16 tileAddress = tileMap + currPixelRow + currPixelCol;
+
+    s16 tileNum;
+    if (unsig) {
+      tileNum = (u8)mmu.read(tileAddress);
+    } else {
+      tileNum = (s8)mmu.read(tileAddress);
+    }
+
+    // adjust for unsigned by using 128 size offset
+    u16 tileLoc = tileData;
+    if (unsig) {
+      tileLoc += (tileNum * 16);
+    } else {
+      tileLoc += ((tileNum + 128) * 16);
+    }
+
+     u8 line = yPos % 8;
+     line *= 2; // two bytes of mem per line
+     u8 byte1 = mmu.read(tileLoc + line);
+     u8 byte2 = mmu.read(tileLoc + line + 1);
+
+     int colorBit = xPos % 8;
+     colorBit -= 7;
+     colorBit *= -1;
+
+     int colorId = checkBit(byte2, colorBit);
+     colorId <<= 1;
+     colorId |= checkBit(byte1, colorBit);
 
+     int color = getcolor(colorId, BGP) ;
+     
+     u8 red = 0;
+     u8 green = 0;
+     u8 blue = 0;
+     switch(color) {
+       case 0: red = 155; green = 188 ; blue = 15; break;   // WHITE = 0  #9bbc0f RGB: 155, 188, 15
+       case 1:red = 139; green = 172 ; blue = 15; break;    // LIGHT_GRAY = 1 #8bac0f RGB: 139, 172, 15
+       case 2: red = 48; green = 98 ; blue = 48; break;     // DARK_GRAY = 2  #306230 RGB: 48, 98, 48
+       default: red = 15; green = 56; blue = 15; break;     // BLACK = 3  #0f380f RGB: 15, 56, 15
+     }
+
```

Comment:
If the `frameBuffer` is flat, I think this is the logic for indexing into it. Someone should definitely double-check me though.

```suggestion
u8* pixelStartLocation = frameBuffer + (LCD_WIDTH * 3 * currentLine) + 3 * i;
pixelStartLocation[0] = red;
pixelStartLocation[1] = green;
pixelStartLocation[2] = blue
```